crates/autopilot/src/run_loop.rs

```diff
@@ -400,6 +410,93 @@ impl RunLoop {
         .collect()
     }
 
+    /// Returns true if winning solution is fair or winner is None
+    fn is_solution_fair(
+        winner: Option<&Participant>,
+        remaining: &Vec<Participant>,
+        auction: &domain::Auction,
+    ) -> bool {
+        let Some(winner) = winner else { return true };
+        let Some(fairness_threshold) = winner.driver.fairness_threshold else {
+            return true;
+        };
+
+        // Returns the surplus difference in the buy token if `left`
+        // is better for the trader than `right`, or 0 otherwise.
+        // This takes differently partial fills into account.
+        let improvement_in_buy = |left: &TradedAmounts, right: &TradedAmounts| {
```

Comment:
Not trying to open the PR, but just for completeness and for having a record when/if we implement combinatorial auctions:

As @MartinquaXD mentioned, this function is not comparing surplus but exchange rates, which makes a difference for partially fillable orders.

An example would be two trades for an order selling 2 ETH for at least 4000 USDC. The proposed executions are
- 1 ETH to 2500 USDC
- 2 ETH to 4900 USDC
The first execution has a better price, the second execution has larger surplus.

For the auction in general we choose to use surplus and not exchange rates for ranking. Using exchange rates for ranking leads to smaller traded amounts and AMMs out of equilibrium.

In practice, the implementation in this PR will not make much of a difference I suppose.

---

For a sell order, surplus is `buy - sell * limit_buy / limit_sell`. The difference in surplus is thus
```
(left_buy - right_buy) - (left_sell - right_sell) * limit_buy / limit_sell.
```
The quantity computed in the code is
```
(right_sell * left_buy - left_sell * right_buy) / right_sell
```
which simplifies to `left_buy - left_sell * right_buy / left_sell`. Comparing that to the difference of surplus gives a difference of
```
(left_sell - right_sell) * (righ_buy / right_sell - limit_buy / limit_sell).
```
That term becomes zero for order which are not partially fillable (i.e. `left_sell = right_sell`) or if the right solution is settling at limit price. Otherwise the is a difference which can be positive or negative. The example above shows that even the ranking can be different.

Reply:
This PR only concerns itself with preventing EBBO violations. Are these triggered by suboptimal surplus or suboptimal exchange rates?
I think this should decide by which metric we need to sort order executions here.